MoveAndRename.md

```diff
@@ -0,0 +1,164 @@
+process is slow, error prone (e.g. disk full), and can require re-uploading
+files in some cases (e.g. Google Drive files on ChromeOS).
+
+We propose adding new `FileSystemHandle::rename()` and
+`FileSystemHandle::move()` methods. `rename()` guarantees atomic moves of a file
+or directory without the need to duplicate data. `move()` only offers atomic
+moves of a file or directory if it is moved within the same file system, while
+moves to non-local file systems, will not be guaranteed to be atomic and may
+involve duplicating data. See the [Open Questions](#open-questions---move).
+
+# API Surface
+We propose adding a `rename()` method and a `move()` method with two variations.
+
+`rename()` will rename a file or directory. It is guaranteed to be atomic.
+
+The first `move()` variation allows a user to specify a `destination_directory`
+to move to, which may or may not be on the same file system. The name of the
+original file or directory is retained. This is not guaranteed to be atomic,
+since the destination directory may be on a different file system and/or may be
+subject to other checks (ex: Safe Browsing in Chrome).
+
+The second `move()` variation allows a user to specify a `new_entry_name` as
+well as a `destination_directory` to move to. The same (lack of) atomicity
+guarantees apply here as the first `move()` variation.
+
+```
+interface FileSystemHandle {
+    ...
+
+    Promise<void> rename(USVString new_entry_name);
+    Promise<void> move(FileSystemDirectoryHandle destination_directory);
+    Promise<void> move(FileSystemDirectoryHandle destination_directory, USVString new_entry_name);
+
+    ...
+};
+```
+
+## Open Questions - Naming
+- Should we prefer `move()` to move verbose alternatives, such as `moveTo()` or
+  `moveToDirectory()`?
+- Should we bother guaranteeing (in the spec) atomicity to all handles `move()`d
+  on the same file system, or only for `rename()`d handles?
+
+# Handling Moves to Other Directories
+
+There are three cases to consider when moving to other directories:
+
+## (1) Moving to a Non-Local File System
+Moving across file systems cannot be guaranteed to be atomic if the operation
+cannot be completed successfully (ex: network connection issues, lack disk
+space, etc.). This may result in partial writes to the target directory.
+
+Note that this approach is no worse than what is available currently.
+
+## (2) Moving out of the OPFS
+
+[#310](https://github.com/WICG/file-system-access/pull/310) exposes a strong use
+case for fast, atomic moves to other directories on the same file system. Files
+can be written efficiently in the Origin Private File System using an
+AccessHandle, then efficiently moved to a directory of the user's choice.
+However, writes using an AccessHandle are not subjected to Safe Browsing checks
+in Chrome. Moving files out of the OPFS to a user's directory will require
+running Safe Browsing checks on all moved files.
```

Comment:
Will this also instruct the operating system to stamp the file with a "mark of the web" (e.g. the work behind https://source.chromium.org/chromium/chromium/src/+/main:components/services/quarantine/quarantine.h;drc=0e7c838c357344ffe4afc1b74c500df3c4caeee9;l=70 in Chromium)?

Reply:
Yes. The process for moving files should be the same as `FileSystemAccessFileWriterImpl`:

1) [scan the file](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/file_system_access/file_system_access_file_writer_impl.cc;l=375;bpv=0;bpt=1)
2) [move the file](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/file_system_access/file_system_access_file_writer_impl.cc;l=437;bpv=0;bpt=1)
3) [add"mark of the web"](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/file_system_access/file_system_access_file_writer_impl.cc;l=495;bpv=0;bpt=1)

The eventual spec language here would have to be more general, since "Safe Browsing" is a Chrome-specific concept anyways

Reply:
Ok, great. I'd suggest mentioning that explicitly... I agree that the language could be more general, but mentioning the kinds of security checks that might be necessary seems pretty reasonable to me, and might help others reading this later understand the requirements more clearly.

(Also, "Safe Browsing" isn't Chrome-specific. :) AFAIR, Safari and Firefox both use it, though to varying degrees, and with varying implementations (proxies, etc).)

Reply:
Updated this section to mention MOTW explicitly, and also included a new "Open Question" about how to handle MOTW in same-file-system moves